Feature/display product data

krushna-simform opened the pull request:
# Description

- Implemented functionality for fetching data from API.
- Created templates for product card and product details page.
- Implemented product search functions with debounce.
- Also added Detail page for products.

# Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Documentation update
- [ ] Refactoring
- [ ] Performance improvement
- [ ] Scaffolding

# Screenshots

![Screenshot from 2025-03-28 19-01-50](https://github.com/user-attachments/assets/757c4b75-ad3a-4e1c-ae86-b4f93badf3e1)
![Screenshot from 2025-03-28 19-02-00](https://github.com/user-attachments/assets/673643da-a8b7-4574-b7fe-20c7b3b146ca)
![Screenshot from 2025-03-28 19-11-00](https://github.com/user-attachments/assets/f46c73cf-cf9c-4f3f-8260-45a9fa1cb8fc)


# Validations

![Screenshot from 2025-03-28 19-13-03](https://github.com/user-attachments/assets/8a821fae-ade2-401c-9340-96f6ce93fda7)
![Screenshot from 2025-03-28 19-13-29](https://github.com/user-attachments/assets/f09db62b-1947-446c-91a1-963fdb9c315a)
![Screenshot from 2025-03-28 19-14-39](https://github.com/user-attachments/assets/7f3a4017-a81b-444f-bf9a-a21ae10f79b3)

# How to test

Deploy Preview: https://deploy-preview-3--product-management-system-ts.netlify.app/

krushna-simform committed:
feature: create interface for Product, Review, and response

krushna-simform committed:
feature: create apiServices for provide fetch functionality

krushna-simform committed:
feature: create container for load product cards

krushna-simform committed:
feature: create innerHTML template for product card

krushna-simform committed:
feature: crate ProductView class for render products card

krushna-simform committed:
feature: crate productController class which handle api calls and loading data for product cards

krushna-simform committed:
feature: initalize ProductContrller class inside index.ts

krushna-simform committed:
feature: move model, view and controller folder inside scripts

krushna-simform committed:
feature: add search by product api

krushna-simform committed:
feature: create product.html file with 'product-details' container

krushna-simform committed:
feature: add handler for viewDetails button

krushna-simform committed:
feature: add eventHandler function for handle events

krushna-simform committed:
feature: add innerHTML template for product-detail page

krushna-simform committed:
feature: create ProductDetails class for handle and print product details data

krushna-simform committed:
feature: update index.ts file for create class instance based on url

krushna-simform committed:
feature: update package.json file for run product.html

krushna-simform committed:
feature: add styling for porduct-details page

krushna-simform committed:
feature: remove extra logs

krushna-simform committed:
feature: add search query api

krushna-simform committed:
feature: create debounce function

krushna-simform committed:
feature: add event handler foe searchbar

krushna-simform committed:
feature: add product.html in build command

Sakshi-Simform reviewed (approved):
ui looks good

dhruv-d-simform commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,73 @@
+import { ProductView } from '../view/productView';
+import { ApiServices } from './apiServices';
+
+export class ProductControler {
```
Nit
Did you mean `ProductController` ?
```suggestion
export class ProductController {
```

krushna-simform replied:
Renamed function names (https://github.com/krushna-simform/product-management-system/pull/3/commits/7007557a824ca530bdaa5a53df8427b0966f6a0b)

dhruv-d-simform commented on src/scripts/controller/apiServices.ts:
```diff
@@ -0,0 +1,23 @@
+        return data.products;
+    }
+
+    async searchOneProduct(query: string): Promise<Product> {
```
Nit.  
We can give more meaningful name here.
```suggestion
    async fetchProductById(query: string): Promise<Product> {
```

krushna-simform replied:
Renamed functions name(https://github.com/krushna-simform/product-management-system/pull/3/commits/7007557a824ca530bdaa5a53df8427b0966f6a0b)

dhruv-d-simform commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,73 @@
+        this.initalize();
+    }
+
+    private async initalize(): Promise<void> {
```
Nit.
Did you mean `initialize` ?
```suggestion
    private async initialize(): Promise<void> {
```

krushna-simform replied:
Rename function name(https://github.com/krushna-simform/product-management-system/pull/3/commits/7007557a824ca530bdaa5a53df8427b0966f6a0b)

dhruv-d-simform commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,73 @@
+    private apiServices: ApiServices;
+    private allProducts: Product[] = [];
+    private isLoading: boolean = true;
+    private view!: ProductView;
```
I think we don't need this Definite assignment assertion here. Because the `view` is initialized in constructor. Let me know your thoughts.

krushna-simform replied:
Removed `!` assertion (https://github.com/krushna-simform/product-management-system/pull/3/commits/7007557a824ca530bdaa5a53df8427b0966f6a0b)

dhruv-d-simform commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,73 @@
+        this.apiServices = new ApiServices();
+        this.view = new ProductView();
+
+        this.searchBar = document.getElementById(
+            'search-bar'
+        ) as HTMLInputElement;
```
Is it a good practice to just cast the result without validating it ? Thoughts ?

krushna-simform replied:
Added validation check for `searchBar` (https://github.com/krushna-simform/product-management-system/pull/3/commits/6ae499580ac22feec87b178208096209e697aa37)

dhruv-d-simform commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,73 @@
+            this.view.renderProducts(this.allProducts);
+            this.eventHandler();
+        } catch (err) {
+            console.error('Error fetching products:', err);
```
We can use `alert` here to provide user the information about error.

krushna-simform replied:
Added `alert` (https://github.com/krushna-simform/product-management-system/pull/3/commits/6f94ffd68cf59c1df5d9e2747657f6f85145605c)

dhruv-d-simform commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,73 @@
+            this.allProducts = [...apiProducts];
+            this.view.renderProducts(this.allProducts);
+        } catch (err) {
+            console.error('Search error:', err);
```
We can use alert here to provide user the information about error.

krushna-simform replied:
Added `alert` (https://github.com/krushna-simform/product-management-system/pull/3/commits/6f94ffd68cf59c1df5d9e2747657f6f85145605c)

dhruv-d-simform commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    initSearchProduct(): void {
```
Nit.
We can provide a more meaningful name here. i.e. `attachSearchHandler`

krushna-simform replied:
Changed function name(https://github.com/krushna-simform/product-management-system/pull/3/commits/e0209e794a9f8ef7324e63e4f59fc626081dc9d5)

dhruv-d-simform commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,73 @@
+        });
+    }
+
+    private eventHandler(): void {
```
Nit.
We can provide a more meaningful name here. i.e. `attachEventHandlers`

krushna-simform replied:
Changed function name(https://github.com/krushna-simform/product-management-system/pull/3/commits/e0209e794a9f8ef7324e63e4f59fc626081dc9d5)

dhruvdhaduk-simform commented on src/scripts/view/productDetails.ts:
```diff
@@ -0,0 +1,32 @@
+    constructor() {
+        this.apiServices = new ApiServices();
+        this.urlParams = new URLSearchParams(window.location.search);
+        this.productId = this.urlParams.get('id')!;
```
Is these a way to avoid this non-null assertion here ? Thoughts ?

dhruv-d-simform commented on src/scripts/view/productDetails.ts:
```diff
@@ -0,0 +1,32 @@
+            this.productDetailContainer.innerHTML =
+                templates.productDetails(product);
+        } catch (err) {
+            console.log('Error fetching product details:', err);
```
We can use alert here to provide user the information about error.

krushna-simform replied:
Add `alert` (https://github.com/krushna-simform/product-management-system/pull/3/commits/bd2d7ae4fd8c4ef131a231d2e0cdeaddb2e3e71d)

dhruv-d-simform commented on src/scripts/view/productDetails.ts:
```diff
@@ -0,0 +1,32 @@
+        this.apiServices = new ApiServices();
+        this.urlParams = new URLSearchParams(window.location.search);
+        this.productId = this.urlParams.get('id')!;
+        this.productDetailContainer = document.getElementById(
+            'product-details'
+        ) as HTMLElement;
```
Is it a good practice to just cast the result without validating it ? Thoughts ?

krushna-simform replied:
Added validation check(https://github.com/krushna-simform/product-management-system/pull/3/commits/f017d59a1f03ed30b9aead599fd662c6cd226999)

dhruv-d-simform commented on src/scripts/view/productView.ts:
```diff
@@ -0,0 +1,40 @@
+    private productContainer: HTMLElement;
+
+    constructor() {
+        this.productContainer = document.getElementById(
+            'product-container'
+        ) as HTMLElement;
```
Is it a good practice to just cast the result without validating it ? Thoughts ?

krushna-simform replied:
Added validation check(https://github.com/krushna-simform/product-management-system/pull/3/commits/f017d59a1f03ed30b9aead599fd662c6cd226999)

dhruvdhaduk-simform commented on src/scripts/view/productView.ts:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    handleViewDetailsButton(product: Product[]) {
+        document
+            .querySelectorAll('.view-details-btn')
+            .forEach((btn, index: number) => {
+                btn.addEventListener('click', () => {
+                    const productId: number = product[index].id;
+                    window.location.href = `product.html?id=${productId}`;
+                });
+            });
```
Is it a good practice to use index to find the productID ?  
Can we use `data-id` attributes to store IDs ?

dhruv-d-simform commented on src/scripts/view/template.ts:
```diff
@@ -0,0 +1,105 @@
+                    <p class="product-title">${product.title}</p>
+                    <p class="product-brand">by ${product.brand}</p>
+                    <div class="product-price-des">
+                        <span class="product-detail-price">$${product.price}</span>
```
We can also add it here as you added in `productCard` to ensure consistency ?

```suggestion
                        <span class="product-detail-price">$${product.price.toFixed(2)}</span>
```

krushna-simform replied:
Added `toFix` (https://github.com/krushna-simform/product-management-system/pull/3/commits/b836ad679c5079aa203f2250f83158297ed3f874)

dhruv-d-simform reviewed (commented):
Provided few suggestions. Also, add few comments in code.

krushna-simform committed:
feature: rename function name and remove assignement assertion

krushna-simform committed:
feature: add validation check for searcBar

krushna-simform committed:
feature: add alert instead of log statement

krushna-simform committed:
feature: change function names

krushna-simform committed:
feature: reinitalize new function names

krushna-simform committed:
feature: add alert instead of log statement

krushna-simform committed:
feature: add validation check befor initalize element

krushna-simform committed:
feature: add toFixed for product detail template

Kaushal-SF commented:
@krushna-simform I think you mean `debounce`.

<img width="300" alt="image" src="https://github.com/user-attachments/assets/6fbe0ec8-4ba1-451a-8ba6-984707db6354" />

Kaushal-SF commented:
@krushna-simform We can improve the UX providing a message when there are no items match with search result.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/96258fed-3e92-4d0d-b815-6aa23eff7e3b" />

krushna-simform commented:
> @krushna-simform I think you mean `debounce`.
> 
> <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/97082607/428507901-6fbe0ec8-4ba1-451a-8ba6-984707db6354.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0MDk2NzcsIm5iZiI6MTc0MzQwOTM3NywicGF0aCI6Ii85NzA4MjYwNy80Mjg1MDc5MDEtNmZiZTBlYzgtNGJhMS00NTFhLThiYTYtOTg0NzA3ZGI2MzU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzMxVDA4MjI1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2YmI4ZmZmMTRkMzI5MGNkZmRjNzBiYTczZWFjYzhmNjE1NGQzNGM0YjQ0ZGVmYWJlZTJjOWJhNGY3NzE3MzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5E_ugxjaTpfovyMvZ5cpUEtxlk0aD_LE7CGAHZi929I">

I have Updated PR Template

Kaushal-SF commented:
@krushna-simform I think we can open a details page when we click on the image. It will improve the user experience.
You can change the view details button name to `View Details`.

krushna-simform commented:
> @krushna-simform I think we can open a details page when we click on the image. It will improve the user experience. You can change the view details button name to `View Details`.

I will implement it.

Kaushal-SF commented:
We can improve the UI for relatively smaller screen.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/07350d64-ae1e-4b20-909e-d0af515e7aaa" />

Kaushal-SF commented on package.json:
You can provide a meaningful name instead of just `pms`.

krushna-simform replied:
Provided meaning full names (https://github.com/krushna-simform/product-management-system/pull/3/commits/ce1819d176ba073dc9c8b26b48fc908681c2d9d4)

krushna-simform committed:
feature: implement no product found messege displyed when no product found while searching

Kaushal-SF commented on src/scripts/controller/apiServices.ts:
```diff
@@ -0,0 +1,23 @@
+export class ApiServices {
+    private URL: string = 'https://dummyjson.com/products';
+
+    async fetchProducts(): Promise<Product[]> {
```
I would prefer this syntax for defining array type. If you like it you can change all the occurrences.
Like: `Array<Product>`.

krushna-simform replied:
Added `Array<Product>` (https://github.com/krushna-simform/product-management-system/pull/3/commits/ce1819d176ba073dc9c8b26b48fc908681c2d9d4)

Kaushal-SF replied:
When I say all the occurrences, that means anywhere you have use syntax like, type[] should be change to Array<type>. Have checked all the places?

krushna-simform replied:
Yes I changed for all.

Kaushal-SF replied:
I afraid but I found few in this file `src/scripts/model/interfaces.ts`.

krushna-simform replied:
Again My bad, i will change it.

krushna-simform replied:
Added `Array<>` syntax(https://github.com/krushna-simform/product-management-system/pull/3/commits/148ef5b1c6c207c50b93a416d7de51d1c8e11041)

Kaushal-SF commented on src/scripts/controller/apiServices.ts:
```diff
@@ -0,0 +1,23 @@
+import { Product, ProductResponse } from '../model/interfaces';
```
While importing a types we can use `type` keyword which ensures that TypeScript treats the import as a type-only import, preventing unnecessary runtime imports.

```suggestion
import type { Product, ProductResponse } from '../model/interfaces';
```

krushna-simform replied:
Added `type` keyword (https://github.com/krushna-simform/product-management-system/pull/3/commits/0a20f43723af3fded7c85c3b389bf717aab9364c)

krushna-simform committed:
feature: add addEventListener on image of product card for View Details

Kaushal-SF commented on src/scripts/controller/apiServices.ts:
```diff
@@ -0,0 +1,23 @@
+import { Product, ProductResponse } from '../model/interfaces';
+
+export class ApiServices {
+    private URL: string = 'https://dummyjson.com/products';
```
Minor Nit: can make it `readonly` to ensure no one can modify it.
```suggestion
    private readonly URL: string = 'https://dummyjson.com/products';
```

krushna-simform replied:
Added `readonly` (https://github.com/krushna-simform/product-management-system/pull/3/commits/210e958ec1173b5ae54ea2e536a91d7abaea3d27)

Kaushal-SF commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,75 @@
+import { Product } from '../model/interfaces';
```
```suggestion
import type { Product } from '../model/interfaces';
```

krushna-simform replied:
Added `type` keyword (https://github.com/krushna-simform/product-management-system/pull/3/commits/0a20f43723af3fded7c85c3b389bf717aab9364c)

Kaushal-SF commented on src/scripts/controller/apiServices.ts:
```diff
@@ -0,0 +1,23 @@
+export class ApiServices {
+    private URL: string = 'https://dummyjson.com/products';
+
+    async fetchProducts(): Promise<Product[]> {
+        const response = await fetch(`${this.URL}`);
+        const data: ProductResponse = await response.json();
+        return data.products;
+    }
```
Do we have any reasons why we are not using `try...catch` here, while making API call?

krushna-simform replied:
I am already handling this 

![Screenshot from 2025-03-31 15-32-46](https://github.com/user-attachments/assets/d12c7c58-4520-420f-b0ce-2295268975aa)

Kaushal-SF replied:
Okay. Please make sure that you try to create a scenario when you encounter the error and then check the behavior where fetchData() is handling it properly.

krushna-simform committed:
feature: add media query for more responsiveness

Kaushal-SF commented on src/scripts/controller/apiServices.ts:
```diff
@@ -0,0 +1,23 @@
+        return data.products;
+    }
+
+    async fetchProductById(query: string): Promise<Product> {
```
I think Id should be a type of number. Any reason we have query type as string?

krushna-simform replied:
Changed type for Query (https://github.com/krushna-simform/product-management-system/pull/3/commits/8ea76be2e9477c6544ea41a818012c0d55cc486a)

Kaushal-SF commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,75 @@
+            this.allProducts = [...apiProducts];
+            this.view.renderProducts(this.allProducts);
+            this.attachEventHandlers();
+        } catch (err) {
+            alert(`Get Error while loading product data ${err}`);
+        } finally {
```
We can also add a console.error() when we encounters error. For other instances as well.

Additionally, we can create a one view for showing meaningful error in the UI of something goes wrong.

krushna-simform replied:
Added `console.error()` (https://github.com/krushna-simform/product-management-system/pull/3/commits/543bac7bcacfec6d0ccfb0727b67605437a3ba6c)

Kaushal-SF commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,75 @@
+            this.allProducts = [...apiProducts];
+            this.view.renderProducts(this.allProducts);
+        } catch (err) {
+            alert(`Ger Erroe while searching product data ${err}`);
```
Typo:
```suggestion
            alert(`Get Error while searching product data ${err}`);
```

krushna-simform replied:
Fixed (https://github.com/krushna-simform/product-management-system/pull/3/commits/74e20295f0d2a81466868352eded0588ddae80d5)

krushna-simform commented:
> @krushna-simform We can improve the UX providing a message when there are no items match with search result.
> 
> <img alt="image" width="500" src="https://private-user-images.githubusercontent.com/97082607/428533882-96258fed-3e92-4d0d-b815-6aa23eff7e3b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0MDk5MDEsIm5iZiI6MTc0MzQwOTYwMSwicGF0aCI6Ii85NzA4MjYwNy80Mjg1MzM4ODItOTYyNThmZWQtM2U5Mi00ZDBkLWI4MTUtNmFhMjNlZmY3ZTNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzMxVDA4MjY0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxOTZkMzE0MDhkM2Y1ZmIwMWZiNWMyOGI0NWNlOGFkZWYwNGIwZmFlZDViMmM5NWI1ODM2NzdkYTIzYmZjYTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SJU09NYi8px8vAox_abFiv-oSYvGaTjJev-8L7ymurk">

Implemented that (https://github.com/krushna-simform/product-management-system/pull/3/commits/a28a60d6902701f213ae7820c1f9d78527c073e3)

"I had already implemented that, but I forgot to add the `return` keyword.

![Screenshot from 2025-03-31 14-39-27](https://github.com/user-attachments/assets/40384714-fa08-4591-b967-af9819553bf3)

Kaushal-SF commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    async searchData(query: string): Promise<void> {
+        try {
+            const apiProducts = await this.apiServices.searchProducts(query);
```
I tested search input providing an empty string. It still makes an API calls which is redundant. We should cover that scenario for better performance and to limit the API responses.

krushna-simform replied:
Added validation for Empty string (https://github.com/krushna-simform/product-management-system/pull/3/commits/2c6c57e8fd1b57799d7638f913e350057202aca5)

krushna-simform commented:
> @krushna-simform I think we can open a details page when we click on the image. It will improve the user experience. You can change the view details button name to `View Details`.

I have implemented that (https://github.com/krushna-simform/product-management-system/pull/3/commits/121ce7284c9d011972d6ff0218498c329c4426d9)

krushna-simform commented:
> We can improve the UI for relatively smaller screen.
> 
> <img alt="image" width="500" src="https://private-user-images.githubusercontent.com/97082607/428535809-07350d64-ae1e-4b20-909e-d0af515e7aaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0MTAwMzEsIm5iZiI6MTc0MzQwOTczMSwicGF0aCI6Ii85NzA4MjYwNy80Mjg1MzU4MDktMDczNTBkNjQtYWUxZS00YjIwLTkwOWUtZDBhZjUxNWU3YWFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzMxVDA4Mjg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwNzI3YzBjYTMyMjM4NWZhMGVjMjdkYzRjNDgwYzAyZjkwM2E5YjhjMzUxZWM3OGJlMTM4YTQ1MzYwYTU5ZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bjOGZxiebrMTsvtVgHHm5FAOY2Da5-37Fqboen4LUmo">

i have added some media query for more responsiveness (https://github.com/krushna-simform/product-management-system/pull/3/commits/37cb0088f012050122a81b9971d2c36c56c90e24)

Kaushal-SF commented on src/scripts/controller/productController.ts:
```diff
@@ -0,0 +1,75 @@
+
+    async fetchData(): Promise<void> {
+        try {
+            const apiProducts = await this.apiServices.fetchProducts();
+
```
Can we make sure that we are getting the proper products list?

krushna-simform replied:
If am making interface of type for `apiProducts` then i am not able to convert data in `json()` because it is showing that custom interface or type does not have json() .

and we are already checking that for renturn type of function.

![Screenshot from 2025-03-31 16-24-21](https://github.com/user-attachments/assets/5b3db2c3-62cb-44bd-b619-603bd681fcb3)

Kaushal-SF replied:
Okay, no issues.

Kaushal-SF commented:
It breaks he UI when we pass unwanted id numer?

<img width="1007" alt="image" src="https://github.com/user-attachments/assets/aa400124-fa5c-4df9-a183-571e63079bb8" />

Kaushal-SF commented on src/scripts/index.ts:
```diff
@@ -1 +1,18 @@
+import { ProductDetails } from './view/productDetails';
+
+document.addEventListener('DOMContentLoaded', () => {
+    const isProductDetailPage =
+        window.location.pathname.includes('product.html');
```
We can also use a `data-` attribute to HTML page and use it here for verification.
Thoughts?

krushna-simform replied:
Can we keep it as it is?

Kaushal-SF replied:
Yup, totally up to you.

Kaushal-SF commented on src/scripts/model/interfaces.ts:
```diff
@@ -0,0 +1,48 @@
+    brand: string;
+    sku: string;
+    weight: number;
+    dimensions: {
+        width: number;
+        height: number;
+        depth: number;
+    };
```
We can create a separate interface for nested types.
Same for `meta` at line no- 25.

krushna-simform replied:
Created separate interface for nested types (https://github.com/krushna-simform/product-management-system/pull/3/commits/a09fb75fb20ee21e06fa2e728ea73f0e4739af46)

Kaushal-SF commented on src/scripts/utils/debounce.ts:
```diff
@@ -0,0 +1,12 @@
+    func: T,
+    delay: number
+) {
+    let timer: ReturnType<typeof setTimeout> | null = null;
```
```suggestion
    let timer: ReturnType<typeof setTimeout> | undefined;
```

krushna-simform replied:
Fixed ([0c5a0ee](https://github.com/krushna-simform/product-management-system/commit/0c5a0eeb6fd1c2cb53cce045dc1155266754bbab))

Kaushal-SF commented on src/scripts/utils/debounce.ts:
```diff
@@ -0,0 +1,12 @@
+) {
+    let timer: ReturnType<typeof setTimeout> | null = null;
+    return (...args: Parameters<T>) => {
+        if (timer) clearTimeout(timer);
+        timer = setTimeout(() => {
```
Nit:

```suggestion
        if (timer !== undefined) clearTimeout(timer);
        timer = setTimeout(() => {
```

krushna-simform replied:
Fixed ([0c5a0ee](https://github.com/krushna-simform/product-management-system/commit/0c5a0eeb6fd1c2cb53cce045dc1155266754bbab))

Kaushal-SF commented on src/scripts/utils/debounce.ts:
```diff
@@ -0,0 +1,12 @@
+    return (...args: Parameters<T>) => {
+        if (timer) clearTimeout(timer);
+        timer = setTimeout(() => {
+            func(...args);
```
Can attach a `this` along with args.
```suggestion
            func.apply(this, ...args);
```

krushna-simform replied:
this is showing error that this has type `any`

Kaushal-SF replied:
You need to provide a type for it.

krushna-simform replied:
Fixed (https://github.com/krushna-simform/product-management-system/pull/3/commits/0c5a0eeb6fd1c2cb53cce045dc1155266754bbab)

Kaushal-SF replied:
@krushna-simform can we use `ThisParameterType<T>`?
- https://www.typescriptlang.org/docs/handbook/utility-types.html#thisparametertypetype

krushna-simform replied:
Yes, we can i will try to implement that.

Kaushal-SF commented on src/scripts/view/template.ts:
```diff
@@ -0,0 +1,105 @@
+import { Product, Review } from '../model/interfaces';
```
```suggestion
import type { Product, Review } from '../model/interfaces';
```

krushna-simform replied:
Added `type` keyword ([0a20f43](https://github.com/krushna-simform/product-management-system/commit/0a20f43723af3fded7c85c3b389bf717aab9364c))

krushna-simform committed:
feature: add product not found template for bad product id

Kaushal-SF commented on src/scripts/view/template.ts:
```diff
@@ -0,0 +1,105 @@
+                <button class="delete-details-btn" aria-label="Delete product" id="${product.id}">
+                    Delete
+                </button>
+                <button class="edit-details-btn" aria-label="Edit product id="${product.id}">Edit</button>
+                <button class="view-details-btn" aria-label="View Details id="${product.id}">
```
Nit: missed to provide quotes.

```suggestion
                <button class="edit-details-btn" aria-label="Edit product" id="${product.id}">Edit</button>
                <button class="view-details-btn" aria-label="View Details" id="${product.id}">
```

krushna-simform replied:
Fixed (https://github.com/krushna-simform/product-management-system/pull/3/commits/54b9418355391a1820b8c5f3a647dedefe3bd253)

krushna-simform commented:
> It breaks he UI when we pass unwanted id numer?
> 
> <img alt="image" width="1007" src="https://private-user-images.githubusercontent.com/97082607/428553808-aa400124-fa5c-4df9-a183-571e63079bb8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0MTMwNzYsIm5iZiI6MTc0MzQxMjc3NiwicGF0aCI6Ii85NzA4MjYwNy80Mjg1NTM4MDgtYWE0MDAxMjQtZmE1Yy00ZGY5LWExODMtNTcxZTYzMDc5YmI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzMxVDA5MTkzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2NjFiMjI3MTI1Y2IwZTUyYzJlMmI0N2ExYWIwNGUyNGM3ZGM3YTYyMTkwNmNhNGVhMDhlNmNmYTlhOWU5NmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uttjWSSj5AD4ZOncQ1aYGu2Phspz7LUch7UsjSB5YjU">

Added template for bad request (https://github.com/krushna-simform/product-management-system/pull/3/commits/7aba6fe0bb36e501eef810a6fe17bd60c315116f)

![Screenshot from 2025-03-31 15-03-14](https://github.com/user-attachments/assets/e9222cd9-7fee-45f0-ab09-0ad1090709bb)

Kaushal-SF commented on src/scripts/view/template.ts:
```diff
@@ -0,0 +1,105 @@
+                    <li><strong>Returns: </strong>Returns ${product.returnPolicy}</li>
+                </ul>
+            </div>
+            ${
+                product.reviews.length > 0
+                    ? `
+            <div class="reviews-section">
+                <p>Customer Reviews</p>
+                <div class="reviews">
+                    ${product.reviews
+                        .map(
+                            (review: Review) => `
+                    <div class="review">
+                        <div class="review-header">
+                        <div class="reviewer">${review.reviewerName}</div>
+                        <div class="review-rating">${'★'.repeat(review.rating)}</div>
+                        <div class="review-date">${new Date(review.date).toLocaleDateString()}</div>
+                        </div>
+                        <p class="review-comment">${review.comment}</p>
+                    </div>
+                    `
+                        )
+                        .join('')}
+                </div>
+            </div>
+            `
+                    : ''
+            }
```
We can extract it into separate function for better readability and maintainability.

krushna-simform replied:
Added another function for review (https://github.com/krushna-simform/product-management-system/pull/3/commits/90833198d8880cbd6702f0e439ad7a08e65a140a)

Kaushal-SF commented on src/scripts/view/productDetails.ts:
```diff
@@ -0,0 +1,38 @@
+    constructor() {
+        this.apiServices = new ApiServices();
+        this.urlParams = new URLSearchParams(window.location.search);
+        this.productId = this.urlParams.get('id')!;
```
This will return null in case if Id is not available.

We can verify that ID is available before moving ahead.
```suggestion
        this.productId = this.urlParams.get('id') || '';
        
       if(!this.productId){
       // handle error
       }
```

krushna-simform replied:
Fixed (https://github.com/krushna-simform/product-management-system/pull/3/commits/90833198d8880cbd6702f0e439ad7a08e65a140a)

Kaushal-SF replied:
I think we didn't check what happened in case if we encounter no `productId`?

krushna-simform replied:
My bad, i will implement that.

krushna-simform replied:
Added validation check for `productId` (https://github.com/krushna-simform/product-management-system/pull/3/commits/ead9ecf3d136405bef8d8439211c15d569bdd3e2)

Kaushal-SF reviewed (commented):
@krushna-simform Wonderful effort for using MVC pattern. Appreciate it.

However, I found few places where we can improve our code for better readability and maintainability.

krushna-simform committed:
feature: add Array<Product> instead of Product[]

krushna-simform committed:
feature: add type keyword for importing types

krushna-simform committed:
feature: add readonly to api URL

krushna-simform committed:
feature: add number as a type for query

krushna-simform committed:
feature: add console.error

krushna-simform committed:
feature: add validation for empty string in search bar

krushna-simform committed:
feature: created separate inreface for nested types

krushna-simform committed:
feature: add missed quotes

krushna-simform committed:
feature: create another function template for review section

krushna-simform committed:
feature: reconfigure typo

krushna-simform committed:
feature: add Array<> syntax in interface.ts

krushna-simform committed:
feature: add validatoin check for productID

krushna-simform committed:
feature: reconfigure debounce function

Kaushal-SF reviewed (approved):
Added one final comment. 

Other than that, code looks good.
